File: server_python/network/protocol.py
"""
NetEase UX RPC Protocol Codec
Implements binary serialization and deserialization matching NetEase's UX.RPC / Lua RPC layer.
"""

import struct
from io import BytesIO
from typing import Any, List, Optional, Tuple

# Packet Modes defined in RPCCommonDefine.lua
MODE_INVOKE = 1
MODE_RETURN = 2
MODE_NOTIFY = 3

MARK_NULL = 0
MARK_COMMON = 255


class BinaryWriter:
    """Implements .NET-style BinaryWriter with 7-bit encoded integers and UX RPC primitives."""

    def __init__(self):
        self.buffer = bytearray()

    def get_bytes(self) -> bytes:
        return bytes(self.buffer)

    def write_byte(self, val: int):
        self.buffer.append(val & 0xFF)

    def write_boolean(self, val: bool):
        self.buffer.append(1 if val else 0)

    def write_int16(self, val: int):
        self.buffer.extend(struct.pack("<h", val))

    def write_uint16(self, val: int):
        self.buffer.extend(struct.pack("<H", val))

    def write_int32(self, val: int):
        self.buffer.extend(struct.pack("<i", val))

    def write_uint32(self, val: int):
        self.buffer.extend(struct.pack("<I", val))

    def write_int64(self, val: int):
        self.buffer.extend(struct.pack("<q", val))

    def write_uint64(self, val: int):
        self.buffer.extend(struct.pack("<Q", val))

    def write_single(self, val: float):
        self.buffer.extend(struct.pack("<f", val))

    def write_double(self, val: float):
        self.buffer.extend(struct.pack("<d", val))

    def write_7bit_encoded_int(self, val: int):
        v = val & 0xFFFFFFFF
        while v >= 0x80:
            self.buffer.append((v | 0x80) & 0xFF)
            v >>= 7
        self.buffer.append(v & 0xFF)

    def write_string(self, val: Optional[str], nullable: bool = False):
        if val is None:
            if nullable:
                self.write_byte(MARK_NULL)
            else:
                self.write_7bit_encoded_int(0)
            return

        if nullable:
            self.write_byte(MARK_COMMON)

        encoded = val.encode("utf-8")
        self.write_7bit_encoded_int(len(encoded))
        self.buffer.extend(encoded)

    def write_list_7bit(self, items: list, writer_func):
        self.write_7bit_encoded_int(len(items))
        for item in items:
            writer_func(self, item)


class BinaryReader:
    """Implements .NET-style BinaryReader with 7-bit encoded integers."""

    def __init__(self, data: bytes):
        self.stream = BytesIO(data)

    @property
    def remaining(self) -> int:
        pos = self.stream.tell()
        self.stream.seek(0, 2)
        end = self.stream.tell()
        self.stream.seek(pos)
        return end - pos

    def read_byte(self) -> int:
        b = self.stream.read(1)
        if not b:
            raise EOFError("Unexpected end of stream")
        return b[0]

    def read_boolean(self) -> bool:
        return self.read_byte() != 0

    def read_int16(self) -> int:
        data = self.stream.read(2)
        if len(data) < 2:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<h", data)[0]

    def read_uint16(self) -> int:
        data = self.stream.read(2)
        if len(data) < 2:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<H", data)[0]

    def read_int32(self) -> int:
        data = self.stream.read(4)
        if len(data) < 4:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<i", data)[0]

    def read_uint32(self) -> int:
        data = self.stream.read(4)
        if len(data) < 4:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<I", data)[0]

    def read_int64(self) -> int:
        data = self.stream.read(8)
        if len(data) < 8:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<q", data)[0]

    def read_uint64(self) -> int:
        data = self.stream.read(8)
        if len(data) < 8:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<Q", data)[0]

    def read_single(self) -> float:
        data = self.stream.read(4)
        if len(data) < 4:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<f", data)[0]

    def read_double(self) -> float:
        data = self.stream.read(8)
        if len(data) < 8:
            raise EOFError("Unexpected end of stream")
        return struct.unpack("<d", data)[0]

    def read_7bit_encoded_int(self) -> int:
        count = 0
        shift = 0
        while shift < 35:
            b = self.read_byte()
            count |= (b & 0x7F) << shift
            shift += 7
            if (b & 0x80) == 0:
                return count
        raise ValueError("Invalid 7-bit encoded int format")

    def read_string(self, nullable: bool = False) -> Optional[str]:
        if nullable:
            mark = self.read_byte()
            if mark == MARK_NULL:
                return None

        length = self.read_7bit_encoded_int()
        if length == 0:
            return ""
        data = self.stream.read(length)
        return data.decode("utf-8", errors="ignore")

    def read_bytes(self, length: int) -> bytes:
        return self.stream.read(length)


class RpcPacket:
    """Represents a UX RPC Packet."""

    def __init__(
        self,
        mode: int,
        method_id: int,
        invoke_id: int = 0,
        err_id: int = 0,
        payload: bytes = b"",
    ):
        self.mode = mode
        self.method_id = method_id
        self.invoke_id = invoke_id
        self.err_id = err_id
        self.payload = payload

    def encode(self, include_framing: bool = True) -> bytes:
        """Encodes packet with TCP 4-byte Int32 length prefix."""
        body = bytearray()
        body.append(self.mode & 0xFF)
        body.extend(struct.pack("<i", self.method_id))

        if self.mode in (MODE_INVOKE, MODE_RETURN):
            body.extend(struct.pack("<i", self.invoke_id))

        if self.mode == MODE_RETURN:
            body.extend(struct.pack("<i", self.err_id))

        body.extend(self.payload)

        if include_framing:
            header = struct.pack("<I", len(body))
            return header + bytes(body)
        return bytes(body)

    @classmethod
    def decode(cls, data: bytes) -> "RpcPacket":
        """Decodes raw packet data (without length prefix)."""
        if len(data) < 5:
            raise ValueError(f"Packet too short: {len(data)} bytes")

        mode = data[0]
        method_id = struct.unpack_from("<i", data, 1)[0]
        offset = 5
        invoke_id = 0
        err_id = 0

        if mode in (MODE_INVOKE, MODE_RETURN):
            if len(data) < offset + 4:
                raise ValueError("Packet truncated at invokeId")
            invoke_id = struct.unpack_from("<i", data, offset)[0]
            offset += 4

        if mode == MODE_RETURN:
            if len(data) < offset + 4:
                raise ValueError("Packet truncated at errId")
            err_id = struct.unpack_from("<i", data, offset)[0]
            offset += 4

        payload = data[offset:]
        return cls(mode, method_id, invoke_id, err_id, payload)

File: server_rust/src/protocol.rs
pub const MODE_INVOKE: u8 = 1;
pub const MODE_RETURN: u8 = 2;
pub const MODE_NOTIFY: u8 = 3;

pub const MARK_NULL: u8 = 0;
pub const MARK_COMMON: u8 = 255;

pub struct BinaryWriter {
    pub buffer: Vec<u8>,
}

impl BinaryWriter {
    pub fn new() -> Self {
        Self { buffer: Vec::new() }
    }

    pub fn write_byte(&mut self, val: u8) {
        self.buffer.push(val);
    }

    pub fn write_int32(&mut self, val: i32) {
        self.buffer.extend_from_slice(&val.to_le_bytes());
    }

    pub fn write_uint32(&mut self, val: u32) {
        self.buffer.extend_from_slice(&val.to_le_bytes());
    }

    pub fn write_uint64(&mut self, val: u64) {
        self.buffer.extend_from_slice(&val.to_le_bytes());
    }

    pub fn write_double(&mut self, val: f64) {
        self.buffer.extend_from_slice(&val.to_le_bytes());
    }

    pub fn write_7bit_encoded_int(&mut self, mut val: u32) {
        while val >= 0x80 {
            self.buffer.push(((val | 0x80) & 0xFF) as u8);
            val >>= 7;
        }
        self.buffer.push((val & 0xFF) as u8);
    }

    pub fn write_string(&mut self, val: &str, nullable: bool) {
        if nullable {
            self.write_byte(MARK_COMMON);
        }
        let bytes = val.as_bytes();
        self.write_7bit_encoded_int(bytes.len() as u32);
        self.buffer.extend_from_slice(bytes);
    }

    pub fn into_bytes(self) -> Vec<u8> {
        self.buffer
    }
}

pub struct BinaryReader<'a> {
    pub data: &'a [u8],
    pub offset: usize,
}

impl<'a> BinaryReader<'a> {
    pub fn new(data: &'a [u8]) -> Self {
        Self { data, offset: 0 }
    }

    pub fn remaining(&self) -> usize {
        self.data.len().saturating_sub(self.offset)
    }

    pub fn read_byte(&mut self) -> Result<u8, &'static str> {
        if self.offset >= self.data.len() {
            return Err("Unexpected EOF");
        }
        let b = self.data[self.offset];
        self.offset += 1;
        Ok(b)
    }

    pub fn read_int32(&mut self) -> Result<i32, &'static str> {
        if self.remaining() < 4 {
            return Err("Unexpected EOF");
        }
        let val = i32::from_le_bytes(self.data[self.offset..self.offset + 4].try_into().unwrap());
        self.offset += 4;
        Ok(val)
    }

    pub fn read_uint64(&mut self) -> Result<u64, &'static str> {
        if self.remaining() < 8 {
            return Err("Unexpected EOF");
        }
        let val = u64::from_le_bytes(self.data[self.offset..self.offset + 8].try_into().unwrap());
        self.offset += 8;
        Ok(val)
    }

    pub fn read_7bit_encoded_int(&mut self) -> Result<u32, &'static str> {
        let mut count = 0u32;
        let mut shift = 0;
        while shift < 35 {
            let b = self.read_byte()?;
            count |= ((b & 0x7F) as u32) << shift;
            shift += 7;
            if (b & 0x80) == 0 {
                return Ok(count);
            }
        }
        Err("Invalid 7-bit encoded int")
    }

    pub fn read_string(&mut self, nullable: bool) -> Result<String, &'static str> {
        if nullable {
            let mark = self.read_byte()?;
            if mark == MARK_NULL {
                return Ok(String::new());
            }
        }

        let len = self.read_7bit_encoded_int()? as usize;
        if len == 0 {
            return Ok(String::new());
        }
        if self.remaining() < len {
            return Err("String length out of bounds");
        }
        let s = String::from_utf8_lossy(&self.data[self.offset..self.offset + len]).to_string();
        self.offset += len;
        Ok(s)
    }
}

pub struct RpcPacket {
    pub mode: u8,
    pub method_id: i32,
    pub invoke_id: i32,
    pub err_id: i32,
    pub payload: Vec<u8>,
}

impl RpcPacket {
    pub fn new(mode: u8, method_id: i32, invoke_id: i32, err_id: i32, payload: Vec<u8>) -> Self {
        Self {
            mode,
            method_id,
            invoke_id,
            err_id,
            payload,
        }
    }

    pub fn encode(&self, include_framing: bool) -> Vec<u8> {
        let mut body = BinaryWriter::new();
        body.write_byte(self.mode);
        body.write_int32(self.method_id);

        if self.mode == MODE_INVOKE || self.mode == MODE_RETURN {
            body.write_int32(self.invoke_id);
        }

        if self.mode == MODE_RETURN {
            body.write_int32(self.err_id);
        }

        body.buffer.extend_from_slice(&self.payload);

        let body_bytes = body.into_bytes();
        if !include_framing {
            return body_bytes;
        }

        let mut framed = BinaryWriter::new();
        framed.write_uint32(body_bytes.len() as u32);
        framed.buffer.extend_from_slice(&body_bytes);
        framed.into_bytes()
    }

    pub fn decode(data: &[u8]) -> Result<Self, &'static str> {
        if data.len() < 5 {
            return Err("Packet too short");
        }

        let mut reader = BinaryReader::new(data);
        let mode = reader.read_byte()?;
        let method_id = reader.read_int32()?;
        let mut invoke_id = 0;
        let mut err_id = 0;

        if mode == MODE_INVOKE || mode == MODE_RETURN {
            invoke_id = reader.read_int32()?;
        }

        if mode == MODE_RETURN {
            err_id = reader.read_int32()?;
        }

        let payload = reader.data[reader.offset..].to_vec();
        Ok(Self::new(mode, method_id, invoke_id, err_id, payload))
    }
}
